### PR TITLE
Move Persistent Queue to exporthelper/internal [Part 1/3]

### PR DESCRIPTION
**Pull request by pmm-sumo**
This addresses one of the items raised in #4025 (move persistent_queue to internal). The overall refactor is broken into three parts:

* **Part 1**: extracting ProducerConsumerQueue interface out of BoundedQueue (so it could be later used for both memory-based and disk-based queues, preparing ground for Part 2) 👈    (this PR)
* **Part 2**: actual moving Persistent Queue into exporthelper/internal (this also requires exporting `internal.RequestUnmarshaller` and `internal.PersistentRequest`) [(link to draft PR)](https://github.com/open-telemetry/opentelemetry-collector/pull/4116)
* **Part 3**: including metrics on number of batches handled through persistent queue [(link to draft PR)](https://github.com/open-telemetry/opentelemetry-collector/pull/4117)

Link to tracking Issue: #4025

Testing: Unit tests updated, manual tests to follow

### CODE REVIEW

**Comment by pmm-sumo** [comment]
@bogdandrutu I closed https://github.com/open-telemetry/opentelemetry-collector/pull/4068 in favor of three smaller PRs. I think that this could be assigned to @tigrannajaryan similarly as the previous one

File: exporter/exporterhelper/internal/producer_consumer_queue.go [hunk 1]
```diff
@@ -1,32 +1,45 @@
 // Copyright The OpenTelemetry Authors
+// Copyright (c) 2019 The Jaeger Authors.
```
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/producer_consumer_queue.go:2]: Why this change?

File: exporter/exporterhelper/internal/producer_consumer_queue.go [hunk 2]
```diff
@@ -1,32 +1,45 @@
...
+// Copyright (c) 2017 Uber Technologies, Inc.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.
 // You may obtain a copy of the License at
 //
-//       http://www.apache.org/licenses/LICENSE-2.0
+// http://www.apache.org/licenses/LICENSE-2.0
```
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/producer_consumer_queue.go:9]: Why this?

File: exporter/exporterhelper/internal/producer_consumer_queue.go [hunk 3]
```diff
@@ -1,32 +1,45 @@
...
+
+// ProducerConsumerQueue defines a producer-consumer exchange which can be backed by e.g. the memory-based ring buffer queue
+// (boundedMemoryQueue) or via a disk-based queue (persistentQueue)
+type ProducerConsumerQueue interface {
 	// StartConsumers starts a given number of goroutines consuming items from the queue
 	// and passing them into the consumer callback.
 	StartConsumers(num int, callback func(item interface{}))
+	// StartConsumersWithFactory creates a given number of consumers consuming items
+	// from the queue in separate goroutines.
+	StartConsumersWithFactory(num int, factory func() Consumer)
```
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/producer_consumer_queue.go:32]: Nobody calls this, consider to remove.

File: exporter/exporterhelper/internal/producer_consumer_queue.go [hunk 4]
```diff
@@ -1,32 +1,45 @@
...
 	// Produce is used by the producer to submit new item to the queue. Returns false if the item wasn't added
 	// to the queue due to queue overflow.
 	Produce(item interface{}) bool
+	// Size returns the current Size of the queue
+	Size() int
+	// Capacity returns capacity of the queue
+	Capacity() int
+	// Resize changes the capacity of the queue, returning whether the action was successful
+	Resize(capacity int) bool
```
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/producer_consumer_queue.go:41]: Nobody calls this, consider to remove.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]

**Comment by bogdandrutu** [comment]
Lots of unnecessary funcs added to the interface, strongly encourage to keep always the API at minimum.

**Comment by pmm-sumo** [comment]
> Lots of unnecessary funcs added to the interface, strongly encourage to keep always the API at minimum.

Good point. I was thinking about removing the unused API from the BoundedQueue but wasn't sure about it so left it like it is. I will remove the unused parts

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 1 above)
Comment (pmm-sumo) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:2]: This is derivative of Jaeger's `BoundedQueue` (though after refactoring and breaking down into interface and implementation). Should it be removed then?

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 2 above)
Comment (pmm-sumo) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:9]: That was the template used in `internal` so I adjusted. But I think we should actually use the format present elsewhere (with indentation)

File: exporter/exporterhelper/internal/producer_consumer_queue.go [hunk 5]
```diff
@@ -12,21 +14,27 @@
...
 
-package exporterhelper
+package internal
 
-// consumersQueue is largely based on queue.BoundedQueue and matches the subset used in the collector
-// It describes a producer-consumer exchange which can be backed by e.g. the memory-based ring buffer queue
-// (queue.BoundedQueue) or via disk-based queue (persistentQueue)
-type consumersQueue interface {
+// Consumer consumes data from a bounded queue
+type Consumer interface {
```
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/producer_consumer_queue.go:20]: Is this used anywhere?

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 1 above)
Comment (bogdandrutu) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:2]: The interface was not copied from Jaeger so no need to add the new copyright I believe. Not sure how this works.

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 5 above)
Comment (mx-psi) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:20]: It is used here https://github.com/open-telemetry/opentelemetry-collector/pull/4115/files#diff-e3cf80389040c560d0447ca06b063cbe062967fe24772e2b1d40cd5c696226b7R59-R61 I added that interface to the Jaeger codebase a while ago :)

Comment (pmm-sumo) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:20]: After we move all to "internal" (PR part 2) - we could unexport it, I think

Comment (bogdandrutu) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:20]: Yes, sorry unused outside the package. No other public API exposes it.

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 1 above)
Comment (pmm-sumo) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:2]: Yeah, I wasn't fully confident about it either. @mx-psi what do you think?

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 5 above)
Comment (pmm-sumo) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:20]: Currently it's being used here: https://github.com/pmm-sumo/opentelemetry-collector/blob/pq-move-to-internal-part1/exporter/exporterhelper/persistent_queue.go#L51-L53 - so if that works we could keep it as it is here and I will update the second PR changing `Consumer` to `consumer`

**Review by bogdandrutu — APPROVED** [review]

File: exporter/exporterhelper/internal/producer_consumer_queue.go (hunk 1 above)
Comment (mx-psi) [reply exporter/exporterhelper/internal/producer_consumer_queue.go:2]: I am not really related to the Jaeger project (just happened to contribute to `pkg/queue` a while back) nor do I know enough about copyright law to answer this, sorry 😄
